core/shared/src/main/scala/zio/ZEnvironment.scala

```diff
@@ -99,14 +99,18 @@ final class ZEnvironment[+R] private (
    * service collisions, which may not be reflected in statically known types,
    * the right hand side will be preferred.
    */
-  def unionAll[R1](that: ZEnvironment[R1]): ZEnvironment[R with R1] =
+  def unionAll[R1](that: ZEnvironment[R1]): ZEnvironment[R with R1] = {
+    val (self0, that0) = if (self.index + that.index < self.index) (self.clean, that.clean) else (self, that)
```

Comment:
Should not this be `< Int.Max`?

Reply:
Since `index` should always be greater than or equal to zero if adding `self.index` and `that.index` results in a value that is less than `self.index` then that means we have wrapped around and need to clean.